### PR TITLE
Feat/add automatic rounding

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
Adds automatic rounding mechanism.

1. Go through a graph and finds TLUs subgraph nodes. 
2. Analyze the outputs of those TLUs on the calibrated input range. 
3. Identify the intervals (step sizes) where the TLU is constant and the offset where the intervals start
4. Raises the precision of TLU inputs to map the step sizes to a power of two
5. Modifies the Graph and the TLU to include rounding and cancel out the raised precision in the TLU
6. Applies the previous steps on all TLUs for each unique  dimension of TLU output (like apply_mapped_lookup_table)
7. Find the best rounding configuration by taking the one with lowest MSE reconstruction with respect to the original TLU

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,7 +54,6 @@ setup_env:
 	echo "Finished installing poetry lock."
 
 	echo "Installing $(CONCRETE_PYTHON_VERSION)" && \
-	poetry run python -m pip install -U --pre "$(CONCRETE_PYTHON_VERSION)"
```
Comment (andrei-stoian-zama) [inline Makefile:57]: I have no idea why this ishere

File: conftest.py [hunk 2]
```diff
@@ -223,7 +223,12 @@ def autoseeding_of_everything(request):
     relative_file_path = absolute_path[test_dir_index + 1 :]
 
     # Derive the sub_seed from the randomly_seed and the test name
-    derivation_string = f"{relative_file_path} # {str(request.node.name)} # {randomly_seed}"
+    derivation_string = (
+        f"{relative_file_path} # {str(request.node.name)} "
+        f"# {randomly_seed} # {str(request.node.own_markers)}"
+    )
+
+    print(f"{derivation_string=}")
```
Comment (andrei-stoian-zama) [inline conftest.py:231]: to be removed

File: Makefile [hunk 3]
```diff
@@ -9,6 +9,7 @@ SRC_DIR:=src
 TEST?=tests
 N_CPU?=4
 CONCRETE_PACKAGE_PATH=$(SRC_DIR)/concrete
+SPHINX_APIDOC_EXCLUDE=$(SRC_DIR)/concrete/ml/common/preprocessors.py
```
Comment (RomanBredehoft) [inline Makefile:12]: what is this exactly ?

File: src/concrete/ml/common/preprocessors.py [hunk 4]
```diff
@@ -0,0 +1,999 @@
...
+        RuntimeError: if the subgraph could not be evaluated
+    """
+
+    node_results: Dict[Node, Union[np.bool_, np.integer, np.floating, np.ndarray]] = {}
+    if sorted_nodes is None:
+        sorted_nodes = list(nx.topological_sort(graph.graph))
+
+    for node in sorted_nodes:
+        if node.operation == Operation.Input:
+            # Deepcopy on the fhe.Graph doesn't modify the input node/indices mappings
```
Comment (RomanBredehoft) [inline src/concrete/ml/common/preprocessors.py:139]: not sure what this comment is about

File: src/concrete/ml/common/preprocessors.py [hunk 5]
```diff
@@ -0,0 +1,999 @@
...
+    if lsbs_to_remove <= 0:
+        return a_node
+
+    assert isinstance(a_node.output.dtype, Integer)
+    rounding_kwargs = {
+        "lsbs_to_remove": lsbs_to_remove,
+    }
+    attributes = {}
+    if rounding_function.__name__ == "round_bit_pattern":
+        # These kwargs are not supported atm
```
Comment (RomanBredehoft) [inline src/concrete/ml/common/preprocessors.py:214]: again what is this comment ? should it require a fixme ?

File: src/concrete/ml/common/preprocessors.py [hunk 6]
```diff
@@ -0,0 +1,999 @@
...
+    )
+    rounding_node.properties["final_lsbs_to_remove"] = lsbs_to_remove
+    rounding_node.properties["resulting_bit_width"] = a_node.output.dtype.bit_width - lsbs_to_remove
+    rounding_node.properties["overflow_detected"] = False
+    rounding_node.properties["original_rounded_bit_width"] = a_node.output.dtype.bit_width
+    if rounding_function.__name__ == "round_bit_pattern":
+        rounding_node.properties["overflow_protection"] = overflow_protection
+        rounding_node.properties["exactness"] = exactness
+
+    rounding_node.bounds = a_node.bounds  # Might be over/under-estimated
```
Comment (RomanBredehoft) [inline src/concrete/ml/common/preprocessors.py:236]: again, should this comment be removed ?

File: src/concrete/ml/common/preprocessors.py [hunk 7]
```diff
@@ -0,0 +1,999 @@
+"""Graph pre-processors for automatic rounding."""
```
Comment (RomanBredehoft) [inline src/concrete/ml/common/preprocessors.py:1]: overall looks good but : 
- very few comments in the actual code, which makes the whole file hard to follow (cannot imagine debugging this in the future) 
- some remaining comments that are confusing / seem irrelevant

File: src/concrete/ml/pytest/torch_models.py [hunk 8]
```diff
@@ -1575,3 +1575,90 @@ def forward(self, x):
         x = self.bn1(x)
         x = self.fc1(x)
         return x
+
+
+class TorchAutoRoundingTLUTester(nn.Module):
+    """A small quantized network with Brevitas, trained on make_classification."""
```
Comment (RomanBredehoft) [inline src/concrete/ml/pytest/torch_models.py:1581]: bad copy paste ?

File: src/concrete/ml/quantization/quantized_ops.py [hunk 9]
```diff
@@ -236,116 +236,7 @@ def q_impl(
 
         p = input2_q_values.shape[-2]
 
-        # Remove the manual matrix multiplication when we can handle input precision with rounding
-        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
-        def enc_mul(x, y):
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_ops.py:241]: is this expected ? we no longer need it ?

File: tests/common/test_preprocessors.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        parameter_encryption_statuses={"x": "encrypted"},
+    )
+    circuit_no_optim_no_rounding = compiler.compile(
+        input_set_as_list_of_array,
+    )
+
+    reference = f(input_set)
+
+    if function_name in ["staircase_pot", "staircase"] and execution_number == 0:
+        # TODO: round to 1b or eliminate these TLUs entirely
```
Comment (RomanBredehoft) [inline tests/common/test_preprocessors.py:167]: do we need a fixme here ?

File: tests/common/test_preprocessors.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    if function_name == "identity":
+        assert circuit.mlir == circuit_no_optim_no_rounding.mlir
+        return
+
+    simulated = numpy.array([circuit.simulate(numpy.array([elt])) for elt in input_set])[..., 0]
+    simulated_no_optim_no_rounding = numpy.array(
+        [circuit_no_optim_no_rounding.simulate(numpy.array([elt])) for elt in input_set]
+    )[..., 0]
+
+    graph_res = numpy.array([circuit.graph(numpy.array([elt])) for elt in input_set])[..., 0]
```
Comment (RomanBredehoft) [inline tests/common/test_preprocessors.py:182]: do we still need to test the graph (old VL) ?

File: tests/common/test_preprocessors.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
+"""Unit tests for TLU optimization preprocessors."""
```
Comment (RomanBredehoft) [inline tests/common/test_preprocessors.py:1]: a bit like above, looks great overall but very few (random) comments that makes the file hard to understand (several times I have no clue what is actually going on)

File: tests/parameter_search/test_automatic_rounding.py [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+        and "opt_round_a" in node.properties["attributes"],
+        ordered=True,
+    )
+
+    # A single TLU should be present
+    assert len(tlu_nodes) == 1
+
+    count_reinterpret = 0
+    count_tlu = 0
+    for line in model.quantized_module_.fhe_circuit.mlir.split("\n"):
```
Comment (RomanBredehoft) [inline tests/parameter_search/test_automatic_rounding.py:143]: should this be a function ? either in source (pytest) or in this file, but looks important

File: tests/parameter_search/test_automatic_rounding.py [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+    """
+
+    params = {
+        "module__n_w_bits": n_bits,
+        "module__n_a_bits": n_bits,
+        "module__n_layers": 2,
+        "max_epochs": 10,
+    }
+
+    model = NeuralNetClassifier(**params)
```
Comment (RomanBredehoft) [inline tests/parameter_search/test_automatic_rounding.py:175]: small detail but you could update `instantiate_model_generic` to be able to provide custom params and then use `preamble` , makes test easier to control

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py [hunk 15]
```diff
@@ -5,23 +5,29 @@
...
 CURRENT_DIR = Path(__file__).resolve().parent
 KEYGEN_CACHE_DIR = CURRENT_DIR.joinpath(".keycache")
 
 # Add MPS (for macOS with Apple Silicon or AMD GPUs) support when error is fixed. For now, we
 # observe a decrease in torch's top1 accuracy when using MPS devices
 # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/3953
 DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
-NUM_SAMPLES = int(os.environ.get("NUM_SAMPLES", 1))
+
+NUM_SAMPLES = int(os.environ.get("NUM_SAMPLES", 1000 if SIMULATE_ONLY else 1))
```
Comment (RomanBredehoft) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py:30]: 1000 samples, even with simulate, looks a bit too much no ? why not keep 100 ?

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py [hunk 16]
```diff
@@ -5,23 +5,29 @@
...
+from concrete.fhe.mlir.processors import ProcessRounding
 from models import cnv_2w2a
 from torch.utils.data import DataLoader
 from trainer import get_test_set
 
+from concrete.ml.common.preprocessors import TLUDeltaBasedOptimizer
 from concrete.ml.quantization import QuantizedModule
 from concrete.ml.torch.compile import compile_brevitas_qat_model
 
+SIMULATE_ONLY = True
```
Comment (RomanBredehoft) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py:21]: should we make this an env var ?

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py [hunk 17]
```diff
@@ -76,11 +82,20 @@ def wrapper(*args, **kwargs):
 # cache generated keys through `insecure_key_cache_location`. As the name suggests, these
 # parameters are unsafe and should only be used for debugging in development
 # Multi-parameter strategy is used in order to speed-up the FHE executions
+base_configuration = Configuration()
```
Comment (RomanBredehoft) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py:85]: to be removed ?

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py [hunk 18]
```diff
@@ -108,12 +124,19 @@ def wrapper(*args, **kwargs):
...
-# Key generation
-print("Creation of the private and evaluation keys.")
-_, keygen_execution_time = measure_execution_time(quantized_numpy_module.fhe_circuit.keygen)(
-    force=True
-)
-print(f"Keygen took {keygen_execution_time} seconds")
+import sys
+
+if sys.platform == "darwin":
+    print("skipping fhe evaluation on darwin platform")
```
Comment (RomanBredehoft) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py:130]: why is that ?

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
thanks for this ! overall it looks great, but we might want to clean a bit some stuff at one point: important files/code are missing soem comments to better explain what is going on (some parts are pretty obscure), several now-irrelevant comments to remove, some things could be simplified in tests

File: Makefile (hunk 3 above)
Comment (andrei-stoian-zama) [reply Makefile:12]: for some reason sphinx was crashing for this file

File: src/concrete/ml/common/preprocessors.py [hunk 19]
```diff
@@ -829,29 +844,43 @@ def apply(self, graph: Graph) -> None:
...
+            #   since we rely on the subgraph to reduce the dimension
+            #   of the TLU.
+            
             # Get TLU sub-graph
-            if "subgraph" not in tlu_node.evaluator.properties["kwargs"]:
+            evaluator = tlu_node.evaluator
+            assert isinstance(evaluator, GenericEvaluator)
+
+            if "subgraph" not in evaluator.properties["kwargs"]:
+                # Not supported for now
```
Comment (andrei-stoian-zama) [inline src/concrete/ml/common/preprocessors.py:859]: when does this case arise ? is it for fhe.univariate use?

Comment (fd0r) [reply src/concrete/ml/common/preprocessors.py:961]: yes univariate or explicit table use in Concrete.

Anyway it's a good check since in the following we try to insert nodes in the subgraph

File: Makefile (hunk 3 above)
Comment (andrei-stoian-zama) [reply Makefile:12]: this can now be removed since we don't have sphinx anymore

File: Makefile [hunk 20]
```diff
@@ -415,7 +416,7 @@ docs_no_links: clean_docs check_docs_dollars
 	mkdir -p docs/_static/
 	@# Generate the auto summary of documentations
 	@# Cannot do without specifying top module currently with sphinx-apidoc
-	poetry run sphinx-apidoc --implicit-namespaces -o docs/_apidoc $(CONCRETE_PACKAGE_PATH)
+	poetry run sphinx-apidoc --implicit-namespaces -o docs/_apidoc $(CONCRETE_PACKAGE_PATH) $(SPHINX_APIDOC_EXCLUDE)
```
Comment (andrei-stoian-zama) [inline Makefile:419]: same here

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py [hunk 21]
```diff
@@ -77,11 +81,19 @@ def wrapper(*args, **kwargs):
...
 # Multi-parameter strategy is used in order to speed-up the FHE executions
+base_configuration = Configuration()
+
+tlu_optimizer = TLUDeltaBasedOptimizer()
 configuration = Configuration(
     dump_artifacts_on_unexpected_failures=False,
     enable_unsafe_features=True,
     use_insecure_key_cache=True,
     insecure_key_cache_location=KEYGEN_CACHE_DIR,
+    additional_pre_processors=[
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_one_example_fhe.py:92]: why do we need this? we should implement `compile_torch_model` with `rounding_threshold_bits: { n_bits=AUTO, method = approximate } `

File: use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py [hunk 22]
```diff
@@ -103,9 +105,15 @@ def main(args):
     model.eval()
 
     # Multi-parameter strategy is used in order to speed-up the FHE executions
+    tlu_optimizer = TLUDeltaBasedOptimizer(
+        internal_bit_width_target=24, exactness=fhe.Exactness.EXACT
+    )
     cfg = Configuration(
         verbose=True,
         show_optimizer=args.show_optimizer,
+        additional_pre_processors=[
```
Comment (andrei-stoian-zama) [inline use_case_examples/cifar/cifar_brevitas_training/evaluate_torch_cml.py:114]: see above comment about the right API

**Comment by fd0r** [comment]
Should we maybe close this one for now @andrei-stoian-zama ?

**Comment by andrei-stoian-zama** [comment]
Despite many efforts, the the results were not good enough and rounding to 6 is the better approach
